Bottom: 4cc44fca8f0014d48b531c9cc0834c9618490654
Top:    ecb3209b15b70c7be298f4b7368dea9097b68f74
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-08-02 13:19:02 +0800

tuntap: flush xdp map in the caller

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 3f23e2c..d7499a9 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1626,7 +1626,6 @@ static u32 tun_do_xdp(struct tun_struct *tun,
 	switch (act) {
 	case XDP_REDIRECT:
 		*err = xdp_do_redirect(tun->dev, xdp, xdp_prog);
-		xdp_do_flush_map();
 		if (*err)
 			break;
 		goto out;
@@ -1715,6 +1714,9 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 		act = tun_do_xdp(tun, tfile, xdp_prog, &xdp, &err);
 		if (err)
 			goto err_xdp;
+
+		if (act == XDP_REDIRECT)
+			xdp_do_flush_map();
 		if (act != XDP_PASS)
 			goto out;
